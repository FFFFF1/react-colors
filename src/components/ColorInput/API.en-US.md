| Property       | Description           | Type             | Default       |
|------------|----------------|------------------|--------------|
| model      | picker model | enum: 'rgb.r' 'rgb.g' 'rgb.b' 'hsv.h' 'hsv.s' 'hsv.v' 'lab.l' 'lab.a' 'lab.b' 'cmyk.c' 'cmyk.m' 'cmyk.y' 'cmyk.k' 'hex' 'alpha'|  -  |
| onChange   | a callback function  | function    | -           |
| color      | from ColorPicker | array    | -        |
| min | minimum value | number  | -         |
| max | maximum value | number  | -         |
| step | offset value when click `up` or `down` | number  | -         |
| fixed | precision  | number  | -         |
| scale | scale | number  | -         |
| label | left label | node | - |
| rightLabel | right label | node | - |
| style  | container style | object | - |
| labelStyle  | left label style | object | - |
| inputStyle  | input style | object | - |
| rightLabelStyle  | right label style | object | - |
| sharp | '#' for hex | boolean | false |

