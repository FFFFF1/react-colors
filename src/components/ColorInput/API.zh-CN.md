| 参数       | 说明           | 类型             | 默认值       |
|------------|----------------|------------------|--------------|
| model      | 拾色模式 | enum: 'rgb.r' 'rgb.g' 'rgb.b' 'hsv.h' 'hsv.s' 'hsv.v' 'lab.l' 'lab.a' 'lab.b' 'cmyk.c' 'cmyk.m' 'cmyk.y' 'cmyk.k'|  无  |
| onChange   | 变化回调  | function    | 由`ColorPicker`传入       |
| color      | 制定颜色 | array    | 由`ColorPicker`传入       |
| min | 最小值  | number  | 见表        |
| max | 最大值  | number  | 见表        |
| step | 每次点击`up` `down`所加减的值  | number  | 见表         |
| fixed | 精度  | number  | 见表         |
| scale | 缩放比例  | number  | 1         |
| label | 左侧标签 | node | 见表 |
| rightLabel | 右侧标签 | node | 无 |
| style  | 容器样式 | object | 无 |
| labelStyle  | 左侧标签样式 | object | 无 |
| inputStyle  | 输入框样式 | object | 无 |
| rightLabelStyle  | 右侧标签样式 | object | 无 |
| sharp | hex 的'#'标签 | boolean | false |

默认值：

| 模式  | min | max | step | fixed | label |
|-------|-----|-----|------|-------|-------|
| rgb.r |  0  | 255 |  1   |   0   |   R   |
| rgb.g |  0  | 255 |  1   |   0   |   G   |
| rgb.b |  0  | 255 |  1   |   0   |   B   |
| hsv.h |  0  | 360 |  1   |   0   |   H   |
| hsv.s |  0  |  1  | 0.01 |   2   |   S   |
| hsv.v |  0  |  1  | 0.01 |   2   |   V   |
| lab.l |  0  | 100 |  1   |   0   |   L   |
| lab.a |-128 | 127 |  1   |   0   |   a   |
| lab.b |-128 | 127 |  1   |   0   |   b   |
|cmyk.c |  0  |  1  | 0.01 |   2   |   C   |
|cmyk.m |  0  |  1  | 0.01 |   2   |   M   |
|cmyk.y |  0  |  1  | 0.01 |   2   |   Y   |
|cmyk.k |  0  |  1  | 0.01 |   2   |   K   |
| hex   |000000|FFFFFF| 1  |   0   |  HEX  |
| alpha |  0  |  1  | 0.01 |   2   |   A   |