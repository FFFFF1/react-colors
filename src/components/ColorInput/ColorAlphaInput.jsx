

import React, { Component, PureComponent, PropTypes } from 'react'
import NumberInput from './NumberInput.jsx'

// todo:设置小数点属性

export class ColorAlphaInput extends (PureComponent || Component) {
  static propTypes = {
    color: PropTypes.array,
    model: PropTypes.oneOf(['alpha']),
    value: PropTypes.number,
    fixed: PropTypes.number,
    step: PropTypes.number,
    min: PropTypes.number,
    max: PropTypes.number
  }

  static defaultProps = {
    label: <span>#</span>,
    step: 0.01
  }

  constructor(props) {
    super()
    this.state = {
      value: props.color[3],
      blurValue: props.color[3],
    }
  }

  componentWillMount() {
    this.componentWillReceiveProps(this.props)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ 
      value: nextProps.color[3]
    })
  }

  handleChange = (value, e) => {
    let {color} = this.props

    this.props.onChange([color[0],color[1],color[2],value], e)
    this.setState({ value: value })
  }

  render() {
    // todo 后部标签
    let {styles, label, rightLabel, placeholder} = this.props

    return (
      <div style={ styles.root }>
        <div style={ styles.label }>
          { label }
        </div>
        <NumberInput
          min={ 0 }
          max={ 1 }
          step={ 0.01 }
          fixed={ 2 }
          style={ styles.input }
          value={ this.state.value }
          onChange={ (value, e) => this.handleChange(value, e) }
          placeholder={ placeholder }
        />
        <div style={ styles.rightLabel }>
          { rightLabel }
        </div>
      </div>
    )
  }
}

export default ColorAlphaInput