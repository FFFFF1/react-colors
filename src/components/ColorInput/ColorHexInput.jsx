/* 颜色属性输入 */

// todo: 添加propTypes和defaultProps
// todo: debug拖动改数功能
// todo: 客制化标签
// todo: 配置小数点位数

import React, { Component, PureComponent, PropTypes } from 'react'
import reactCSS from 'reactcss'

export class ColorInput extends (PureComponent || Component) {
  static propTypes = {
    color: PropTypes.array,
    model: PropTypes.oneOf(['hex','css']),
  }

  static defaultProps = {
  }

  constructor(props) {
    super()
    this.state = {
      value: this.getValue(props),
      blurValue: this.getValue(props),
    }
  }

  render() {
    let {label, rightLabel} = this.props
    let value = this.state.value.toFixed(this.getAttr('fixed'))
    let labelNode = label || <span style={ styles.label }>{ this.getAttr('label') }</span>
    return (
      <div style={ styles.root }>
        { labelNode }
        <input
          style={ styles.input }
          ref="input"
          value={ value }
          onKeyDown={ this.handleKeyDown }
          onChange={ this.handleChange }
          onBlur={ this.handleBlur }
          placeholder={ this.props.placeholder }
        />   
      </div>
    )
  }

}

export default ColorInput