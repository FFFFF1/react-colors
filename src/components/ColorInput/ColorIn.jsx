// todo 输入hex或字符串时，渲染一个字符串输入组件，输入input的，渲染另一个数字输入组件