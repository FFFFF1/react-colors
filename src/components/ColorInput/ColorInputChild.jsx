/* 颜色属性输入 */

// todo: 添加propTypes和defaultProps
// todo: debug拖动改数功能
// todo: 客制化标签
// todo: 配置小数点位数
// todo: Lab值变化bug
// todo: fix cmyk.k
// 
// todo: 使用pos 0 1 2 3代替 getColor 和 getValue

// 抽离出对象部分

import React, { Component, PureComponent, PropTypes } from 'react'
import NumberInput from './NumberInput.jsx'
import HexInput from './HexInput.jsx'

export class ColorInputChild extends (PureComponent || Component) {
  static propTypes = {
    color: PropTypes.oneOfType([PropTypes.array,PropTypes.string]),
    model: PropTypes.oneOf(['rgb.r','rgb.g','rgb.b','hsv.h','hsv.s','hsv.v','lab.l','lab.a','lab.b','cmyk.c','cmyk.m','cmyk.y','cmyk.k','alpha','hex']),
    value: PropTypes.number,
    fixed: PropTypes.number,
    step: PropTypes.number,
    min: PropTypes.number,
    max: PropTypes.number,
    scale: PropTypes.number
  }

  static defaultProps = {
    scale: 1
  }

  constructor(props) {
    super()
    this.state = {
      value: null,
    }
  }

  getAttr (attr) {
    let {model} = this.props
    let map = {min:0,max:1,step:2,fixed:3,pos:4}
    let attrs = {
      'rgb.r': [0, 255, 1, 0, 0],
      'rgb.g': [0, 255, 1, 0, 1],
      'rgb.b': [0, 255, 1, 0, 2],
      'hsv.h': [0, 360, 1, 0, 0],
      'hsv.s': [0, 1, 0.01, 2, 1],
      'hsv.v': [0, 1, 0.01, 2, 2],
      'lab.l': [0, 100, 1, 0, 0],
      'lab.a': [-128, 127, 1, 0, 1],
      'lab.b': [-128, 127, 1, 0, 2],
      'cmyk.c': [0, 1, 0.01, 2, 0],
      'cmyk.m': [0, 1, 0.01, 2, 1],
      'cmyk.y': [0, 1, 0.01, 2, 2],
      'cmyk.k': [0, 1, 0.01, 2, 3],
      'alpha': [0, 1, 0.01, 2, 3],
      'hex': [0,16777215,1,null, null]
    }
    return this.props[attr] || attrs[model][ map[attr] ]
  }

  componentWillMount() {
    this.componentWillReceiveProps(this.props)
  }

  componentWillReceiveProps(nextProps) {
    let pos = this.getAttr('pos')
    this.setState({ 
      value: typeof pos === 'number' ? nextProps.color[this.getAttr('pos')] : nextProps.color
    })
  }

  handleChange = (value, e) => {
    let {color} = this.props
    let newColor = color
    let pos = this.getAttr('pos')
    if(typeof pos === 'number') newColor[pos] = value 
    else newColor = value
    this.props.onChange(newColor, e)
    this.setState({ value: value })
  }

  render() {
    let {style,model,fixed,scale,sharp,placeholder} = this.props 
    if( model === 'hex') {
      return (
        <HexInput
          step={ this.getAttr('step') }
          style={ style }
          sharp={ sharp }
          value={ this.state.value }
          onChange={ this.handleChange }
          placeholder={ placeholder }
        />
      )
    }

    if( !fixed ){
      fixed = parseInt(this.getAttr('fixed') - Math.log10(scale))
      fixed < 0 ? 0 : fixed
    } 

    return (
     <NumberInput
        min={ this.getAttr('min') }
        max={ this.getAttr('max') }
        step={ this.getAttr('step') }
        fixed={ fixed }
        style={ style }
        value={ this.state.value }
        scale={ scale }
        onChange={ this.handleChange }
        placeholder={ placeholder }
      />   
    )
  }
}

export default ColorInputChild

/*
  handleDrag = (e) => {
    if (this.props.dragLabel) {
      const newValue = Math.round(this.props.value + e.movementX)
      if (newValue >= 0 && newValue <= this.props.dragMax) {
        this.props.onChange({ [this.props.label]: newValue }, e)
      }
    }
  }

  handleMouseDown = (e) => {
    if (this.props.dragLabel) {
      e.preventDefault()
      this.handleDrag(e)
      window.addEventListener('mousemove', this.handleDrag)
      window.addEventListener('mouseup', this.handleMouseUp)
    }
  }

  handleMouseUp = () => {
    this.unbindEventListeners()
  }

  unbindEventListeners = () => {
    window.removeEventListener('mousemove', this.handleDrag)
    window.removeEventListener('mouseup', this.handleMouseUp)
  }*/
