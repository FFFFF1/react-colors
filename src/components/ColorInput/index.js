export {default} from './ColorInput.jsx'