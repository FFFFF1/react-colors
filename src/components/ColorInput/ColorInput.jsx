/* 颜色属性输入 */

// todo: 添加propTypes和defaultProps
// todo: debug拖动改数功能
// todo: 客制化标签
// todo: 配置小数点位数
// todo: Lab值变化bug
// todo: fix cmyk.k

import React, { Component, PureComponent, PropTypes } from 'react'
import ColorInputChild from './ColorInputChild.jsx'

export class ColorInput extends (PureComponent || Component) {
  static propTypes = {
    model: PropTypes.oneOf(['rgb.r','rgb.g','rgb.b','hsv.h','hsv.s','hsv.v','lab.l','lab.a','lab.b','cmyk.c','cmyk.m','cmyk.y','cmyk.k','hex','alpha']),
    style: PropTypes.object,
    label: PropTypes.node,
    rightLabel: PropTypes.node,
    labelStyle: PropTypes.object,
    rightLabelStyle: PropTypes.object,
    inputStyle: PropTypes.object,
  }

  getDefaultLabel(model) {
    return {
      'rgb.r':'R',
      'rgb.g':'G',
      'rgb.b': 'B',
      'hsv.h': 'H',
      'hsv.s': 'S',
      'hsv.v':'v',
      'lab.l':'L',
      'lab.a':'a',
      'lab.b':'b',
      'cmyk.c': 'C',
      'cmyk.m': 'M',
      'cmyk.y':'Y',
      'cmyk.k': 'K',
      'alpha': 'a',
      'hex': '#'
    }[model]
  }

  render () {
    let {
      style, 
      label,
      rightLabel,
      labelStyle, 
      rightLabelStyle, 
      inputStyle, 
      ...props
    } = this.props

    const styles = {
      root: Object.assign({
        display: 'inline-block',
        position: 'relative',
        marginBottom: 4,
      }, style),
      label: Object.assign({
        display: 'inline-block',
        height: '100%',
        fontSize: 13,
        verticalAlign: 'top',
        width: 20,
        lineHeight: '20px'
      }, labelStyle),
      rightLabel: Object.assign({
        display: 'inline-block',
        height: '100%',
        fontSize: 13,
        verticalAlign: 'top',
        paddingLeft: 4,
        lineHeight: '20px'
      }, rightLabelStyle),
      input: Object.assign({
        verticalAlign: 'top',
        width: 40,
        boxSizing: 'border-box',
        height: 20,
        padding: '0 2px',
        fontSize: 12
      }, inputStyle)
    }

    return (
      <div style={ styles.root }>
        <span style={ styles.label }>
          { label || this.getDefaultLabel(props.model) }
        </span>
        <ColorInputChild style={styles.input} {...props} />
        <span style={ styles.rightLabel }>
          { rightLabel }
        </span>
      </div>
      
    )
  }
}

export default ColorInput