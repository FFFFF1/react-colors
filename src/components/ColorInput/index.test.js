import React from 'react'
import ColorInput from './ColorInput.jsx'
import ColorHexInput from './ColorHexInput.jsx'
import { mount, render } from 'enzyme'
import { renderToJson } from 'enzyme-to-json'

describe('ColorHexInput', () => {
	it('render correctly', () => {
		const wrapper = render(<ColorHexInput color='#2196F3' model='hex'/>)
		expect(renderToJson(wrapper)).toMatchSnapshot()
	})
})

describe('ColorInput', () => {
	it('render correctly', () => {
		const wrapper = render(<ColorInput color={[255,0,0]} model='rgb.r'/>)
		expect(renderToJson(wrapper)).toMatchSnapshot()
	})

	it('render correctly', () => {
		const wrapper = render(<ColorInput color={[255,0,0]} model='rgb.r'/>)
		expect(renderToJson(wrapper)).toMatchSnapshot()
	})

	it('show color correctly', () => {
		const input = mount(<ColorInput color={[123,0,0]} model='rgb.r'/>).find('input')
		expect(input.getNode().value).toBe('123')
	})

	describe('value change set color', () => {
		it('chang input', () => {
			const input = mount(<ColorInput color={[255,0,0]} model='rgb.r' onChange={newColor => {
				expect(newColor).toEqual([123,0,0])
			}}/>).find('input')

			input.simulate('change',{
				target: {
					value: 123
				}
			})
		})
		
		it('key down DOWN', () => {
			const input = mount(<ColorInput color={[123,0,0]} model='rgb.r' onChange={newColor => {
				expect(newColor).toEqual([122,0,0])
			}}/>).find('input')

			input.simulate('keyDown',{
				keyCode: 40
			})
		})
		
		it('key down UP', () => {
			const input = mount(<ColorInput color={[123,0,0]} model='rgb.r' onChange={newColor => {
				expect(newColor).toEqual([124,0,0])
			}}/>).find('input')

			input.simulate('keyUp',{
				keyCode: 38
			})
		})
	})
})