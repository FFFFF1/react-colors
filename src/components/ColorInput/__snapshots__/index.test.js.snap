// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorInput renders correctly render correctly 1`] = `
<div
  style="position:relative;margin-bottom:4px;height:36px;display:flex;align-items:center;"
>
  <span
    style="position:relative;font-size:16px;text-align:center;display:inline-block;width:30px;"
  >
    R
  </span>
  <input
    style="box-sizing:border-box;height:100%;"
    value="255"
  />
</div>
`;
